core/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java

```diff
@@ -142,72 +283,150 @@ public TermsQueryBuilder disableCoord(boolean disableCoord) {
     }
 
+    public String lookupIndex() {
+        return this.termsLookup.index;
+    }
+
     /**
      * Sets the index type to lookup the terms from.
      */
     public TermsQueryBuilder lookupType(String lookupType) {
-        this.lookupType = lookupType;
-        return this;
+        if (lookupType != null) {
+            if (this.termsLookup == null) {
+                this.termsLookup = new TermsLookupBuilder();
+            }
+            this.termsLookup.type = lookupType;
+        }
+        return this;    }
+
+    public String lookupType() {
+        return this.termsLookup.type;
     }
 
     /**
      * Sets the doc id to lookup the terms from.
      */
     public TermsQueryBuilder lookupId(String lookupId) {
-        this.lookupId = lookupId;
+        if (lookupId != null) {
+            if (this.termsLookup == null) {
+                this.termsLookup = new TermsLookupBuilder();
+            }
+            this.termsLookup.id = lookupId;
+        }
         return this;
     }
 
+    public String lookupId() {
+        return this.termsLookup.id;
+    }
+
     /**
      * Sets the path within the document to lookup the terms from.
      */
     public TermsQueryBuilder lookupPath(String lookupPath) {
-        this.lookupPath = lookupPath;
-        return this;
+        if (lookupPath != null) {
+            if (this.termsLookup == null) {
+                this.termsLookup = new TermsLookupBuilder();
+            }
+            this.termsLookup.path = lookupPath;
+        }
+        return this;    }
+
+    public String lookupPath() {
+        return this.termsLookup.path;
     }
 
     public TermsQueryBuilder lookupRouting(String lookupRouting) {
-        this.lookupRouting = lookupRouting;
+        if (lookupRouting != null) {
+            if (this.termsLookup == null) {
+                this.termsLookup = new TermsLookupBuilder();
+            }
+            this.termsLookup.routing = lookupRouting;
+        }
         return this;
     }
 
+    public String lookupRouting() {
+        return this.termsLookup.routing;
+    }
+
+    /**
+     * Same as {@link #convertToBytesRefIfString} but on Iterable.
+     * @param objs the Iterable of input object
+     * @return the same input or a list of {@link BytesRef} representation if input was a list of type string
+     */
+    private static List<Object> convertToBytesRefListIfStringList(Iterable objs) {
+        if (objs == null) {
+            return null;
+        }
+        List<Object> newObjs = new ArrayList<>();
+        for (Object obj : objs) {
+            newObjs.add(convertToBytesRefIfString(obj));
+        }
+        return newObjs;
+    }
+
+    /**
+     * Same as {@link #convertToStringIfBytesRef} but on Iterable.
+     * @param objs the Iterable of input object
+     * @return the same input or a list of utf8 string if input was a list of type {@link BytesRef}
+     */
+    private static List<Object> convertToStringListIfBytesRefList(Iterable objs) {
```

Comment:
I think we shouldn't have this special case for Iterable, as I mentioned above I think it would be good if that constructor delegated to the Objects... constructor and do the potential String->BytesRef conversion there.